harness/determined/_searcher.py

```diff
@@ -0,0 +1,145 @@
+import enum
+import logging
+from typing import Iterator, Optional
+
+import determined as det
+from determined.experimental import client
+
+log = logging.getLogger("generic")
+
+
+class Unit(enum.Enum):
+    EPOCHS = "UNIT_EPOCHS"
+    RECORDS = "UNIT_RECORDS"
+    BATCHES = "UNIT_BATCHES"
+
+
+class SearcherOp:
+    def __init__(
+        self,
+        session: client.Session,
+        trial_id: int,
+        unit: Unit,
+        length: int,
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._unit = unit
+        self._length = length
+        self._completed = False
+
+    @property
+    def unit(self) -> Unit:
+        return self._unit
+
+    @property
+    def length(self) -> int:
+        return self._length
+
+    @property
+    def records(self) -> int:
+        if self._unit != Unit.RECORDS:
+            raise AssertionError(
+                "you can only read op.records if you configured your searcher in terms of records"
+            )
+        return self._length
+
+    @property
+    def batches(self) -> int:
+        if self._unit != Unit.BATCHES:
+            raise AssertionError(
+                "you can only read op.batches if you configured your searcher in terms of batches"
+            )
+        return self._length
+
+    @property
+    def epochs(self) -> int:
+        if self._unit != Unit.EPOCHS:
+            raise AssertionError(
+                "you can only read op.epochs if you configured your searcher in terms of epochs"
+            )
+        return self._length
+
+    def report_progress(self, length: float) -> None:
+        if self._completed and length != self._length:
+            raise AssertionError("you must not call op.report_progress() after op.complete()")
+        log.debug(f"op.report_progress({length})")
+        self._session.post(
+            f"/api/v1/trials/{self._trial_id}/progress",
+            encoded_body=det.util.json_encode(length),
+        )
+
+    def complete(self, searcher_metric: float) -> None:
+
+        if self._completed:
+            raise AssertionError("you may only call op.complete() once")
+        self._completed = True
+        body = {
+            "op": {
+                "length": {
+                    "length": self._length,
+                    "unit": self._unit.value,
+                }
+            },
+            "searcherMetric": searcher_metric,
+        }
+        log.debug(f"op.complete({searcher_metric})")
```

Comment:
I actually think it's much more useful to log all of the generic API calls, because when you know what the API call is, you can know more intelligently how to log it.  Then some day, when a user's Generic API code is misbehaving, we can ask them to turn on the `determined.generic` logger and it will be easy and readable for both us and users, as to exactly how they are using the generic API.  Also, when you log too much or in a hard-to-read format, it makes it harder for users to debug their own code.

Reply:
ok